bsum.1
.TH BSUM 1 BLAKESUM
.SH NAME
bsum - Compute and check BLAKE message digests
.SH SYNOPSIS
.B bsum
[-l bits] [-S salt] [-c | -B | -L | -U] [-xz]
.RI [ file "] ..."
.SH DESCRIPTION
Print or check BLAKE checksums.
.SH OPTIONS
The
.B bsum
utility conforms to the Base Definitions volume of POSIX.1-2017,
.IR "Section 12.2" ,
.IR "Utility Syntax Guidelines" .
.PP
The following options are supported:
.TP
.B -B
Output checksums in binary representation. This suppresses
the filenames and checksum delimiters. Only the checksums
are printed.
.TP
.B -c
Read BLAKE sums from the file and check them against the
files on your system. The input files should be
formatted as the output of this program, or similarly.
This is not going to work if any of the filenames in the
input files starts with <space> or <tab>, or if they
contain a <newline>, unless the
.B -z
option is also used.
.TP
.B -L
Output checksums in lower-case hexadecimal representation. (Default)
.TP
.BI "-l " bits
Select version of the BLAKE algorithm. Valid values
are 224 (default), 256, 384, and 512.
.TP
.BI "-S " salt
Specify a 16-byte (for the 224 and 256 bit versions) or
32-byte salt (for the 384 and 512 bit versions) that the
BLAKE algorithm shall use. This salt shall be expressed
in full length hexadecimal: 32 (for 16-byte) or 64
(for 32-byte) hexadecimal digits, or rather 16 or 32 pairs
of hexadecimal digits, ordered from left to right to specify
the values from index 0 to the last byte. In each pair,
the left digit stores the high bits of the byte, and the
right digit stores the low bits of the byte. For example,
the digit pair
.B 80
represents a byte with the decimal value 128, whereas the
digit pair
.B 08
represents a byte with the decimal value 8. The pairs
are joined without any delimiters, and no byte may be
omitted. So, for a 16 byte salt where each byte's value
is its index, the salt is expressed as
.BR 000102030405060708090a0b0c0d0e0f ,
however each letter may be either small or capital.
If no salt is specified, an all-zeroes salt is used.
.TP
.B -U
Output checksums in upper-case hexadecimal representation.
.TP
.B -x
Convert input files from hexadecimal form to binary form
before calculating the checksums.
.TP
.B -z
Lines end with NUL instead of LF. If used with
.BR -c ,
this applies to read files (not the output), but it will
also apply more strict parsing and allow any whitespace
in file names.
.SH OPERANDS
The following operands are supported:
.TP
.I file
File to read. The standard input will be used
.B -
or no
.I file
is specified.
.SH EXIT STATUS
.TP
0
Successful completion.
.TP
1
Checksums did not match or a file did not exist.
.TP
2
An error occurred.
.SH SEE ALSO
.BR b2sum (1),
.BR b224sum (1),
.BR b256sum (1),
.BR b384sum (1),
.BR b512sum (1)


bsum.c
/* See LICENSE file for copyright and license details. */
#include "common.h"

const char *argv0 = "bsum";

static int lenght_by_command_name = 0;

static int flag_check = 0;
static int flag_binary = 0;
static int flag_lower = 0;
static int flag_upper = 0;
static int flag_hex = 0;
static int flag_zero = 0;
static int length;
static void *flag_salt = NULL;

static void
usage(void)
{
	fprintf(stderr, "usage: %s%s [-S salt] [-c | -B | -L | -U] [-xz] [file] ...\n",
	        argv0, lenght_by_command_name ? "" : " [-l bits]");
	exit(2);
}

static void
get_lenght_by_command_name(const char *command)
{
	const char *p;
	p = strrchr(command, '/');
	p = p ? &p[1] : command;
	if (strstr(p, "b224sum"))
		lenght_by_command_name = 224;
	else if (strstr(p, "b256sum"))
		lenght_by_command_name = 256;
	else if (strstr(p, "b384sum"))
		lenght_by_command_name = 384;
	else if (strstr(p, "b512sum"))
		lenght_by_command_name = 512;
}

static int
hash_fd_blake(int fd, const char *fname, int decode_hex, unsigned char hash[], void *state, void *salt,
              void (*init_func)(void *state, void *salt), size_t (*update_func)(void *state, const void *msg, size_t n),
              size_t (*get_buf_size_func)(size_t bytes, size_t bits, const char *suffix),
              void (*digest_func)(void *state, void *msg, size_t bytes, size_t bits, const char *suffix, unsigned char out[]))
{
	char *buf = NULL;
	size_t size = 0;
	size_t len = 0;
	size_t off = 0;
	size_t req;
	ssize_t r;
	int ok;
	init_func(state, salt);
	for (;;) {
		if (len == size)
			buf = erealloc(buf, size += 8 << 10);
		r = read(fd, &buf[len], size - len);
		if (r <= 0) {
			if (!r)
				break;
			if (errno == EINTR)
				continue;
			fprintf(stderr, "%s: %s: %s\n", argv0, fname, strerror(errno));
			free(buf);
			return -1;
		}
		len += (size_t)r;
		if (!decode_hex) {
			off += update_func(state, &buf[off], len - off);
			if (off == len)
				off = 0;
		}
	}
	if (off)
		memmove(&buf[0], &buf[off], len -= off);
	if (decode_hex) {
		len = libblake_decode_hex(buf, len, buf, &ok);
		if (!ok) {
			fprintf(stderr, "%s: %s: %s\n", argv0, fname, "invalid hexadecimal input");
			free(buf);
			return -1;
		}
	}
	req = get_buf_size_func(len, 0, NULL);
	if (req > size)
		buf = erealloc(buf, size);
	digest_func(state, buf, len, 0, NULL, hash);
	free(buf);
	return 0;
}

static int
hash_fd_blake224(int fd, const char *fname, int decode_hex, unsigned char hash[], void *salt)
{
	struct libblake_blake224_state state;
	return hash_fd_blake(fd, fname, decode_hex, hash, &state, salt,
	                     (void (*)(void *, void *))&libblake_blake224_init2,
	                     (size_t (*)(void *, const void *, size_t))&libblake_blake224_update,
	                     &libblake_blake224_digest_get_required_input_size,
	                     (void (*)(void *, void *, size_t, size_t, const char *, unsigned char[]))&libblake_blake224_digest);
}

static int
hash_fd_blake256(int fd, const char *fname, int decode_hex, unsigned char hash[], void *salt)
{
	struct libblake_blake256_state state;
	return hash_fd_blake(fd, fname, decode_hex, hash, &state, salt,
	                     (void (*)(void *, void *))&libblake_blake256_init2,
	                     (size_t (*)(void *, const void *, size_t))&libblake_blake256_update,
	                     &libblake_blake256_digest_get_required_input_size,
	                     (void (*)(void *, void *, size_t, size_t, const char *, unsigned char[]))&libblake_blake256_digest);
}

static int
hash_fd_blake384(int fd, const char *fname, int decode_hex, unsigned char hash[], void *salt)
{
	struct libblake_blake384_state state;
	return hash_fd_blake(fd, fname, decode_hex, hash, &state, salt,
	                     (void (*)(void *, void *))&libblake_blake384_init2,
	                     (size_t (*)(void *, const void *, size_t))&libblake_blake384_update,
	                     &libblake_blake384_digest_get_required_input_size,
	                     (void (*)(void *, void *, size_t, size_t, const char *, unsigned char[]))&libblake_blake384_digest);
}

static int
hash_fd_blake512(int fd, const char *fname, int decode_hex, unsigned char hash[], void *salt)
{
	struct libblake_blake512_state state;
	return hash_fd_blake(fd, fname, decode_hex, hash, &state, salt,
	                     (void (*)(void *, void *))&libblake_blake512_init2,
	                     (size_t (*)(void *, const void *, size_t))&libblake_blake512_update,
	                     &libblake_blake512_digest_get_required_input_size,
	                     (void (*)(void *, void *, size_t, size_t, const char *, unsigned char[]))&libblake_blake512_digest);
}

int
hash_fd(int fd, const char *fname, int decode_hex, unsigned char hash[])
{
	int ret;

	if (length == 224)
		ret = hash_fd_blake224(fd, fname, decode_hex, hash, flag_salt);
	else if (length == 256)
		ret = hash_fd_blake256(fd, fname, decode_hex, hash, flag_salt);
	else if (length == 384)
		ret = hash_fd_blake384(fd, fname, decode_hex, hash, flag_salt);
	else if (length == 512)
		ret = hash_fd_blake512(fd, fname, decode_hex, hash, flag_salt);
	else
		abort();

	return ret;
}

int
main(int argc, char *argv[])
{
	const char *salt_str = NULL;
	uint_least8_t salt_buf[32];
	int status = 0;
	int output_case;
	char newline;

	if (argv[0])
		get_lenght_by_command_name(argv[0]);

	length = lenght_by_command_name;

	ARGBEGIN {
	case 'c':
		flag_check = 1;
		break;
	case 'B':
		flag_binary = 1;
		break;
	case 'L':
		flag_lower = 1;
		flag_upper = 0;
		break;
	case 'U':
		flag_upper = 1;
		flag_lower = 0;
		break;
	case 'S':
		if (salt_str)
			usage();
		salt_str = ARG();
		break;
	case 'x':
		flag_hex = 1;
		break;
	case 'z':
		flag_zero = 1;
		break;
	case 'l':
		if (length)
			usage();
		length = atoi(ARG());
		if (length != 224 && length != 256 && length != 384 && length != 512) {
			fprintf(stderr, "%s: valid arguments for -l are 224 (default), 256, 384, and 512\n", argv0);
			return 2;
		}
		break;
	default:
		usage();
	} ARGEND;

	if (flag_check + flag_binary + flag_lower + flag_upper > 1)
		usage();

	if (!length)
		length = 224;

	if (salt_str) {
		parse_salt(salt_buf, salt_str, length <= 256 ? 16 : 32);
		flag_salt = salt_buf;
	}

	newline = flag_zero ? '\0' : '\n';
	if (flag_check) {
		if (!argc) {
			status |= -check_and_print("-", (size_t)length, flag_hex, newline);
		} else {
			for (; *argv; argv++)
				status |= -check_and_print(*argv, (size_t)length, flag_hex, newline);
		}
	} else {
		output_case = flag_binary ? -1 : flag_upper;
		if (!argc) {
			status |= -hash_and_print("-", (size_t)length, flag_hex, newline, output_case);
		} else {
			for (; *argv; argv++)
				status |= -hash_and_print(*argv, (size_t)length, flag_hex, newline, output_case);
		}
	}

	if (fflush(stdout) || ferror(stdout) || fclose(stdout)) {
		fprintf(stderr, "%s: %s\n", argv0, strerror(errno));
		return 2;
	}
	return status;
}
